File: search-service/src/utils/tracing.ts
import { NodeSDK } from '@opentelemetry/sdk-node';
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-http';
import { Resource } from '@opentelemetry/resources';
import { SemanticResourceAttributes } from '@opentelemetry/semantic-conventions';
import { SimpleSpanProcessor, ConsoleSpanExporter, BatchSpanProcessor } from '@opentelemetry/sdk-trace-node';
import { HttpInstrumentation } from '@opentelemetry/instrumentation-http';
import { ExpressInstrumentation } from '@opentelemetry/instrumentation-express'; // Assuming Express
// Import Elasticsearch instrumentation if direct ES client is used
// import { ElasticsearchInstrumentation } from '@opentelemetry/instrumentation-elasticsearch';
import { logger } from '../shared/utils/logger'; // Import shared logger


// Fallback logger
// const consoleLogger = { // Removed consoleLogger
//   info: (message: string, ...args: any[]) => console.log(`[Search Tracing INFO] ${message}`, ...args),
//   error: (message: string, ...args: any[]) => console.error(`[Search Tracing ERROR] ${message}`, ...args),
// };

export const initTracer = (serviceName: string, jaegerEndpoint: string, nodeEnv: string) => {
  const traceExporter = new OTLPTraceExporter({
    url: jaegerEndpoint,
  });

  const spanProcessor = nodeEnv === 'production'
    ? new BatchSpanProcessor(traceExporter)
    : new SimpleSpanProcessor(new ConsoleSpanExporter());

  const sdk = new NodeSDK({
    resource: new Resource({
      [SemanticResourceAttributes.SERVICE_NAME]: serviceName,
      [SemanticResourceAttributes.DEPLOYMENT_ENVIRONMENT]: nodeEnv,
    }),
    spanProcessor: spanProcessor,
    instrumentations: [
      new HttpInstrumentation(),
      new ExpressInstrumentation(),
      // new ElasticsearchInstrumentation(), // Add if using Elasticsearch client directly
    ],
  });

  try {
    sdk.start();
    logger.info(`OpenTelemetry Tracing initialized for service: ${serviceName}, environment: ${nodeEnv}`);
    logger.info(`Jaeger exporter configured with endpoint: ${jaegerEndpoint}`);
    if (nodeEnv !== 'production') {
      logger.info('ConsoleSpanExporter is active for development/debugging.');
    }
  } catch (error) {
    logger.error({ err: error }, `Error initializing OpenTelemetry Tracing for ${serviceName}:`);
  }

  const shutdown = () => {
    sdk.shutdown()
      .then(() => logger.info(`OpenTelemetry Tracing terminated for ${serviceName}`))
      .catch((error) => logger.error({ err: error }, `Error terminating OpenTelemetry Tracing for ${serviceName}`))
      .finally(() => process.exit(0));
  };

  process.on('SIGTERM', shutdown);
  process.on('SIGINT', shutdown);

  return sdk;
};

File: search-service-py/app/utils/tracing.py
from opentelemetry import trace
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor, ConsoleSpanExporter
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter as OTLPHttpSpanExporter
from opentelemetry.sdk.resources import Resource, SERVICE_NAME as RESOURCE_SERVICE_NAME

from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor
from opentelemetry.instrumentation.elasticsearch import ElasticsearchInstrumentor # For Elasticsearch client
from opentelemetry.instrumentation.motor import MotorInstrumentor # If using MongoDB
from opentelemetry.instrumentation.httpx import HTTPXClientInstrumentor

from app.core.config import settings
from app.utils.logger import logger

def init_tracer(service_name: str = settings.OTEL_SERVICE_NAME, jaeger_endpoint: str = settings.JAEGER_ENDPOINT):
    if not settings.ENABLE_TRACING:
        logger.info("OpenTelemetry tracing is disabled for search-service.")
        return

    try:
        resource = Resource(attributes={
            RESOURCE_SERVICE_NAME: service_name,
            "environment": settings.ENVIRONMENT,
        })

        provider = TracerProvider(resource=resource)
        otlp_exporter = OTLPHttpSpanExporter(endpoint=jaeger_endpoint)
        processor = BatchSpanProcessor(otlp_exporter)
        provider.add_span_processor(processor)

        if settings.ENVIRONMENT == "development":
            console_exporter = ConsoleSpanExporter()
            console_processor = BatchSpanProcessor(console_exporter)
            provider.add_span_processor(console_processor)
            logger.debug("ConsoleSpanExporter enabled for OpenTelemetry tracing (search-service).")

        trace.set_tracer_provider(provider)
        logger.info(f"OpenTelemetry tracing initialized for service: {service_name}, exporting to {jaeger_endpoint}")

    except Exception as e:
        logger.error(f"Failed to initialize OpenTelemetry tracer for search-service: {e}", exc_info=True)


def instrument_application_search(app): # Renamed
    """Instruments the FastAPI application and related libraries for OpenTelemetry."""
    if not settings.ENABLE_TRACING:
        return

    try:
        FastAPIInstrumentor.instrument_app(app, tracer_provider=trace.get_tracer_provider())

        # Elasticsearch instrumentation (typically global, call once)
        ElasticsearchInstrumentor().instrument(tracer_provider=trace.get_tracer_provider())

        if settings.MONGODB_URI: # Only instrument Motor if MongoDB is configured and used
            MotorInstrumentor().instrument(tracer_provider=trace.get_tracer_provider())

        HTTPXClientInstrumentor().instrument(tracer_provider=trace.get_tracer_provider())

        logger.info("Search service application and libraries instrumented for OpenTelemetry.")
    except Exception as e:
        logger.error(f"Failed to instrument search service app for OpenTelemetry: {e}", exc_info=True)
